components/script/dom/element.rs

```diff
@@ -436,6 +450,59 @@ impl Element {
                 box_.clone_overflow_y() == overflow_y::computed_value::T::Hidden
         })
     }
+
+    pub fn is_shadow_host(&self) -> bool {
+        self.shadow_root.get().is_some()
+    }
+
+    /// https://dom.spec.whatwg.org/#dom-element-attachshadow
+    /// XXX This is not exposed to web content yet. It is meant to be used
+    ///     for UA widgets only.
+    pub fn attach_shadow(&self, is_ua_widget: IsUserAgentWidget) -> Fallible<DomRoot<ShadowRoot>> {
+        // Step 1.
+        if self.namespace != ns!(html) {
+            return Err(Error::NotSupported);
+        }
+
+        // Step 2.
+        match self.local_name() {
+            &local_name!("article") |
+            &local_name!("aside") |
+            &local_name!("blockquote") |
+            &local_name!("body") |
+            &local_name!("div") |
+            &local_name!("footer") |
+            &local_name!("h1") |
+            &local_name!("h2") |
+            &local_name!("h3") |
+            &local_name!("h4") |
+            &local_name!("h5") |
+            &local_name!("h6") |
+            &local_name!("header") |
+            &local_name!("main") |
+            &local_name!("nav") |
+            &local_name!("p") |
+            &local_name!("section") |
+            &local_name!("span") => {},
+            &local_name!("video") | &local_name!("audio")
+                if is_ua_widget == IsUserAgentWidget::Yes => {},
+            _ => return Err(Error::NotSupported),
+        };
+
+        // Step 3.
+        if self.is_shadow_host() {
+            return Err(Error::InvalidState);
+        }
+
+        // Steps 4, 5 and 6.
+        let shadow_root = self
+            .shadow_root
+            .or_init(|| ShadowRoot::new(self, &*self.node.owner_doc()));
+
+        self.node.owner_doc().register_shadow_root(&*shadow_root);
```

Comment:
This should only register it if the host is connected (and should probably also set the connected bit on the ShadowRoot, though that's done in Shadowroot::new so fine).